compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AMD64MacroAssembler.java

```diff
@@ -337,4 +345,420 @@ private void trigEpilogue(Register dest, AMD64Address tmp) {
+                bind(L_avx2_loop);
+                vpmovzxbw(vtmp, new AMD64Address(src, tmp, Scale.Times1));
+                vmovdqu(new AMD64Address(dst, tmp, Scale.Times2), vtmp);
+                addq(tmp, 16);
+                jcc(ConditionFlag.NotZero, L_avx2_loop);
+
+                bind(L_below_threshold);
+                bind(L_avx2_tail);
+
+                movl(tmp, len);
+                andl(tmp, -8);
+                jccb(ConditionFlag.Zero, L_sse_tail);
+                andl(len, 8 - 1);
+
+                // Inflate another 8 bytes before final tail copy.
+                pmovzxbw(vtmp, new AMD64Address(src));
+                movdqu(new AMD64Address(dst), vtmp);
+                addq(src, 8);
+                addq(dst, 16);
+
+                // Fall-through to L_sse_tail.
+            } else {
+                // When there is no AVX2 support available, we use AVX/SSE support to
+                // inflate into maximum 128-bits per operation.
+
+                movl(tmp, len);
+                andl(tmp, -8);
+                jccb(ConditionFlag.Zero, L_sse_tail);
+                andl(len, 8 - 1);
+
+                leaq(src, new AMD64Address(src, tmp, Scale.Times1));
+                leaq(dst, new AMD64Address(dst, tmp, Scale.Times2));
+                negq(tmp);
+
+                Label L_sse_copy_8_loop = new Label();
+                // Inflate 8 bytes (chars) per iteration, reading 64-bit compact vectors
+                // and writing 128-bit inflated ditto.
+                bind(L_sse_copy_8_loop);
+                pmovzxbw(vtmp, new AMD64Address(src, tmp, Scale.Times1));
+                movdqu(new AMD64Address(dst, tmp, Scale.Times2), vtmp);
+                addq(tmp, 8);
+                jcc(ConditionFlag.NotZero, L_sse_copy_8_loop);
+
+                // Fall-through to L_sse_tail.
+            }
+
+            Label L_copy_chars = new Label();
+
+            bind(L_sse_tail);
+            cmpl(len, 4);
+            jccb(ConditionFlag.Less, L_copy_chars);
+
+            movdl(vtmp, new AMD64Address(src));
+            pmovzxbw(vtmp, vtmp);
+            movq(new AMD64Address(dst), vtmp);
+            subq(len, 4);
+            addq(src, 4);
+            addq(dst, 8);
+
+            bind(L_copy_chars);
+        }
+
+        // Inflate any remaining characters (bytes) using a vanilla implementation.
+        testl(len, len);
+        jccb(ConditionFlag.Zero, L_done);
+        leaq(src, new AMD64Address(src, len, Scale.Times1));
+        leaq(dst, new AMD64Address(dst, len, Scale.Times2));
+        negq(len);
+
+        Label L_copy_chars_loop = new Label();
+        // Inflate a single byte (char) per iteration.
+        bind(L_copy_chars_loop);
+        movzbl(tmp, new AMD64Address(src, len, Scale.Times1));
+        movw(new AMD64Address(dst, len, Scale.Times2), tmp);
+        incrementq(len, 1);
+        jcc(ConditionFlag.NotZero, L_copy_chars_loop);
+
+        bind(L_done);
```

Comment:
These assembly snippets are rather large, we might want to transform them into stubs. it can be done later though.
/cc @tkrodriguez

Reply:
Indeed. I was actually considering to make that change (or export them as stubs from HotSpot). But I took a conservative approach to begin with.